## Natural Language Processing with Applications in Economics/Finance
---


  ##### 1. Introduction to NLP
  ##### 2. One-Hot Encoding and Word Embeddings
  ##### 3. Word2vec and GloVe (Global Vectors for Word Representation)
  ##### 4. Recurrent Neural Networks (RNNs) and Long Short-Term Memory (LSTM)
  ##### 5. Attention Mechanisms and Transformers
  ##### 6. BERT and GPTs

### Introduction to Natural Language Processing (NLP)
---

* **Natural Language Processing (NLP)** is a subfield of artificial intelligence that focuses on the interaction between computers and humans through natural language.

* It involves the development of algorithms and models that enable computers to understand, interpret, and generate human language in a way that is valuable.

* **Applications of NLP** include:
  * Sentiment analysis
  * Machine translation
  * Named entity recognition
  * Text summarization
  * Question answering
  * Chatbots and virtual assistants


**Reference**: [Standford CS224N: Natural Language Processing with Deep Learning](https://web.stanford.edu/class/archive/cs/cs224n/cs224n.1246/)

### Meaning of Words for Computers 
---

#### How do we represent the meaning of a word?

**Definition: _meaning_ (Webster Dictionary)**  
- the idea that is **represented by a word**, phrase, etc.  
- the idea that **a person wants to express** by using words, signs, etc.  
- the idea that is expressed in a work of writing, art, etc.

#### Commonest Linguistic View of Meaning

**Signifier (symbol)** ⇔ **Signified (idea or thing)**  
= _Denotational semantics_

**Example:**  
`tree` ⇔ { 🌳, 🌲, 🌴, ... }


### Discrete Representations Don’t Capture Similarity
---
* In traditional NLP, words are regarded as **discrete symbols** and often represented as **one-hot vectors** models.

* **One-hot encoding**: Each word is represented as a vector with a single 1 and all other elements 0.

* **Vector dimension is usually very high!!** (e.g., 100,000+ words in the vocabulary)

> **Example:** User searches for _“Seattle motel”_, but we also want to match _“Seattle hotel”_
```text
    motel = [0 0 0 0 0 0 0 0 1 0 0 0 0]  
    hotel = [0 0 0 0 0 1 0 0 0 0 0 0 0]

### Why WordNet is Not Enough
---
- ✅ __Useful, but lacks nuance__ 
  - e.g., _“proficient”_ is listed as a synonym for _“good”_ — only correct in specific contexts  
  - WordNet includes offensive synonyms without flagging connotations

- ❌ __Fails to capture new meanings or slang__  
  - e.g., _wicked, badass, nifty, wizard, genius, ninja, bombest_  
  - Hard to keep up-to-date!

- ❌ __Subjective and labor-intensive__ 
  - Requires human input to maintain  
  - Cannot compute word similarity reliably


### Discrete Representations Don’t Capture Similarity
---
* In traditional NLP, words are regarded as **discrete symbols** and often represented as **one-hot vectors** models.

* **One-hot encoding**: Each word is represented as a vector with a single 1 and all other elements 0.

* **Vector dimension is usually very high!!** (e.g., 100,000+ words in the vocabulary)

> **Example:** User searches for _“Seattle motel”_, but we also want to match _“Seattle hotel”_
```text
    motel = [0 0 0 0 0 0 0 0 1 0 0 0 0]  
    hotel = [0 0 0 0 0 1 0 0 0 0 0 0 0]

## Text Analysis using One-Hot Encoding/Word Vectors 
---

Based on `Gentzkow et al. (2019, Journal of Economic Literature)`


### 1. What is Text Analysis?

Text analysis refers to a broad set of tools used to extract information from textual data. In recent years, economists have started to use these tools to gain insights from non-traditional data sources.

#### Applications in Economics

- **Finance**: Use of financial news, social media, and company filings to predict stock prices.
- **Macroeconomics**: Text is used to predict inflation, unemployment, and understand policy uncertainty.
- **Consumer Behavior**: Analyzing advertisements and product reviews helps understand decision-making processes.
- **Political Economy**: Politicians' speeches reveal shifts in political agendas over time.

#### Challenge: High Dimensionality

Text data is extremely high-dimensional.

- A document of length $w$ using a vocabulary of $p$ words has $p^w$ possible combinations.
- Example: 30-word Twitter messages using only the 1,000 most common words yield more dimensions than atoms in the universe.



### 2. Text Analysis in a Nutshell

The process generally follows three main steps:

1. **Text Representation**:
   - Convert raw text $\mathcal{D}$ into numerical features $\mathbf{X}$.
   - Reduce dimensionality to something manageable.

2. **Prediction**:
   - Use $\mathbf{X}$ to predict unknown outcomes $\mathbf{Y}$.
   - Apply high-dimensional statistical tools (e.g., classify spam emails).

3. **Inference**:
   - Use predicted values $\hat{\mathbf{Y}}$ in causal or descriptive analysis.





### 3. From Text to Data
---
#### Step 1: Define a Document

A "document" depends on your research question:

- **Spam detection**: One email = one document.
- **Stock prediction**: One day’s financial news = one document.

#### Step 2: Feature Selection

##### a. Cleaning Text
- Remove punctuation, numbers, HTML tags, etc.
- Text Splitting (分词): Break text into words or phrases (tokens).
  - Python libraries like `nltk` or `spaCy` can help.
  - Chinese text may require specialized libraries like [`jieba`](https://github.com/fxsjy/jieba).

##### b. Remove Non-Informative Words
- **Stop words** like "the", "and", "or" add noise. Chinese stop words include "的", "了", "是", etc. [中文停用词表](https://github.com/CharyHong/Stopwords)
- **Very rare words** lack generalizability.


##### c. TF-IDF (Term Frequency-Inverse Document Frequency)

To quantify word importance:

$$
\text{tf}_{ij} = c_{ij} \quad \text{(word count in document $i$)}
$$

$$
\text{idf}_j = \log\left(\frac{n}{\sum_i \mathbf{1}(c_{ij} > 0)}\right)
$$

- **High TF, low DF** → Important
- **Low TF or high DF** → Less informative

##### d. Stemming
- Reduces words to their base form.
  - Example: “economic”, “economics”, “economically” → “economic”

## Application: Measuring Technical Change using Patent Texts
---

### `Kelly et al. (2021, AER Insights)`: Measuring Long-Run Technical Change 

#### 1. Main Contributions

* They use textual analysis to create new indicators of technical change. 

* **Identifying Important Patents (Technologies)**: 
  > **Idea**: Important patent are **distinct** from previous patents, but are **influential** for subsequent patents. 

* **Breakthrough Patents**: the *most important patents*

* Time-series indices for technological change at the aggregate and sectoral level 


#### 2. Data and Methodology 

* **Data Sources**: the USPTO and Google’s patent search engine.
  * The full text of over nine million patents between 1840 and 2010 
  
* **Defining Similarity Metric between Patent Documents**: 
  - For patents: Terms like `electricity` and `petroleum` are more important than the common words `process`, `method` or `inventor`
  
  - Traditional  $\textit{TF-IDF}$: $\text{TFIDF}_{pw}\equiv \text{TF}_{pw} \cdot \text{IDF}_w $
    - **TF**: Term Frequency (how often a term appears in a patent) $$\text{TF}_{pw} = \frac{c_{pw}}{\sum_k c_{pk}}$$
      - where $c_{pw}$ is the count of term $w$ in patent $p$ and $\sum_k c_{pk}$ is the total count of all terms in patent $p$.
      - **TF** captures the relative importance of a term within a specific patent.

    - **IDF**: Inverse Document Frequency (how common a term is across all patents) $$\text{IDF}_w = \log\left(\frac{N}{\sum_p \mathbf{1}(c_{pw} > 0)}\right)$$
      - where $N$ is the total number of patents and $\sum_p \mathbf{1}(c_{pw} > 0)$ is the count of patents containing term $w$.
      - **IDF** captures how unique or common a term is across all patents.
      
  - **Modified TF-IDF for Breakthrough Innovations**: 
      - > For example, Nikola Tesla’s famous 1888 patent of an `AC motor`,  which was among the first patents to use the phrase “`alternating current`,” a phrase used with  great frequency throughout the 20th century
  
      - Define $\textit{backward-}IDF$: $$\text{BIDF}_w = \log \left(\frac{N_\text{ patents prior to }p}{1+ N_{\text{ patents prior to } p  \text{ with term} w}}\right)$$
        - where $p'$ is any patent that comes after patent $p$ in time.
        - However, **Backward-IDF** tends to **downweight** subsequent patents that use the same term, even if they are important. 
          - e.g., GM's patent for `Alternating Current ignition system` in 1990 
        
      - The term frequency are normalized using the term's **initial** $\textit{BIDF}$ value: $$\text{TFBIDF}_{pw} = \text{TF}_{w,i}\cdot \text{BIDF}_{w,t}, t \equiv \min(i,j)$$ 
  
    - Calculate the cosine similarity between the two normalized TF-BIDF vectors $$\rho_{i,j} = \frac{\text{TFBIDF}_{i, t}}{\|\text{TFBIDF}_{i,t}\|}\cdot \frac{\text{TFBIDF}_{j, t}}{\|\text{TFBIDF}_{j,t}\|}$$ 

<div align="center">
<img src="images/pairwise_patent_similarity.png", width="400" />
</div>

  - **Defining Important Patents**:
    - **Idea**: Important patents are those *both novel and impactful* 
  
    - **Novelty**: A patent is novel if it is **dissimilar** to previous patents. $$BS_j^\tau = \sum_{j\in \mathcal{B}_{j,\tau}}\rho_{j,i}$$
  
    - **Impact**: A patent is impactful if it is **similar** to subsequent patents. $$FS_j^\tau = \sum_{j\in \mathcal{F}_{j,\tau}}\rho_{j,i}$$

    - **Patent Importance**: $$\text{q}_j^\tau = \frac{FS_j^\tau}{BS_j^\tau}$$
    - **Breakthrough Patent** are the top 10% important patents. 
  
#### 3. Results and Follow-up Studies

  <div align="center">
  <img src="images/breakthrough_patents.png", width="400" />
  </div>


* [`Kogan et al. (2023, NBER wp)`](https://www.nber.org/papers/w31846) measure the distance between a given technology and the tasks performed in different occupations.   
  * They use TF_IDF as the weights of the words   calculate the cosine similarity between a  breakthrough patent $b$ and *routine* or *non-routine* component of occupation $o$ 
  $$\rho^j_(b,\,o) = \frac{\Chi_b}{\|\Chi_b\|}\cdot \frac{\Chi_o^j}{\Chi_o^j}\quad j \in \{ \text{Routine, Non-routine} \} $$

  * They then use the task-classification based on their capital/labor complementarity to measure workers' exposure to breakthrough technologies. 
  
  <div align="center">
  <img src="images/occupation_technology_exposure.png", width="450" />
  </div>

### 应用拓展 I ： 基于中文专利文本的专利质量测算
---
#### `陈志远等(2025, 工作论文)`: 基于中文专利文本的专利质量测算

* 基于改进和调整的`Kelly et al. (2021)`方法，测算中国`1985-2021`的发明专利质量
  
* **KPST方法可能存在的问题**:
  * 样本期间专利数量激增, 导致对后向相似度计算造成向上的偏误;
  
  * 广泛存在的渐进式创新。特别是在计算前向和后向相似度时，如果不排除相同申请人的专利申请，
可能会导致对这类累积式技术专利质量估计的偏误;

  * 跨技术领域比较 vs. 同一技术领域比较

* **调整后的方法**:$$\text{后向相似度: } \overline{BS}_p(\tau) = \sum_{q\in \mathcal{B}_{p}(\tau) \backslash \mathcal{C}_p(\tau)} \frac{s_{p,q}}{\sum_{k=t_p-\tau}^{t_p-1}N_k} \\\text{前向相似度: } \overline{FS}_p(\tau)=\sum_{q\in \mathcal{F}_p(\tau)\backslash \mathcal{D}_p(\tau)} \frac{s_{p,q}}{\sum_{k=t+1}^{t_p+\tau}N_k} \\ \text{专利质量: } \text{Quality}_p(\tau) = \frac{\overline{FS}_p(\tau)}{\overline{BS}_p(\tau)}$$
  * 其中，$\mathcal{C}_p(\tau)$和$\mathcal{D}_p(\tau)$分别表示专利$p$的同一申请人专利集合和同一技术领域专利集合。

#### 结果
<div align="center">
<b> 不同方法计算专利质量指标相关性</b><br>
<img src="images/不同专利质量指标相关性.png", width="400" />
</div>

<div align="center">
<b> 中外专利质量对比变化</b><br>
<img src="images/中外专利质量对比变化.png", width="400" />
</div>

<div align="center">
<b> 不同发明主体专利质量变化</b><br>
<img src="images/不同主体专利质量与数量.png", width="500" />
</div>

#### Python 程序

参见：jupyter notebook  `patent_text_analysis_notes.ipynb`

### 应用拓展II: 基于文本匹配的知识溢出本地化效应识别
---
#### `陈志远，于皓，张杰(2025, 中国工业经济)`: [中国知识溢出本地化的行政边界效应研究](https://ciejournal.ajcass.com/Magazine/show/?id=117850) 

##### 1. 研究背景 

* 专利之前相互引用倾向于发生在同一个地区内部，但专利之间的引用既反映了可能的知识溢出，也反映了创新知识的地理分布特征：
  * 产业集聚或集群导致知识和技术在空间分布的不均衡
  
* `Jaffe et al. (1993)` 提出通过构造被引专利的相似专利作为对照组来识别知识溢出本地化效应
  * 对于专利对$i \rightarrow j$，如果$i$被$j$引用，则计算$j$的最相似专利$j'$，并将其作为对照组。
  * 若$j'$ 与$j$ 在同一地区，则引用可能是创新集聚
  * 若$j'$ 与$j$ 在不同地区，则引用可能是知识溢出本地化
  
##### 2. 主要贡献

* 利用基于独热法的文本匹配方法，构造专利对之间的相似度指标：
  
  * 中国1985-2018年1300多万条发明专利
  
  * 构建了约500万专利引用组对

* 基于知识溢出本地化的估计结果证实了知识溢出本地化现象，并构建了中国创新知识溢出行政边界效应指数 



## Word Embeddings
---

#### 1. Word Vectors

#### 2. Word2vec

#### 3. GloVe 

### 1. Word Vectors
---
#### Representing Words by Their Context 

* Distributional semantics: A word’s meaning is given by the words that frequently appear close-by

    * *“You shall know a word by the company it keeps”* (J. R. Firth 1957: 11)

    * One of the most successful ideas of modern statistical NLP!

* When a word $w$ appears in a text, its context is the set of words that appear nearby (within a fixed-size window).

* We use the many contexts of $w$ to build up a representation of $w$

    > **Examples with "banking"**:

    > ...government debt problems turning into **banking** crises as happened in 2009...

    > ...saying that Europe needs unified **banking** regulation to replace the hodgepodge...

    > ...India has just given its **banking** system a shot in the arm...

  * The word "banking" appears in different contexts, which can be used to create a vector representation that captures its meaning.

#### Word Embeddings
We will build a dense vector for each word, chosen so that it is similar to vectors of words that appear in similar contexts, measuring similarity as the vector dot (scalar) product.
\begin{align*}
\text{\color{blue} banking}= \begin{bmatrix}
0.286 \\ 
0.792  \\
-0.177  \\
-0.107  \\
0.109  \\
-0.542  \\
0.349  \\
0.271\\
\end{bmatrix}\quad &\quad  \text{\color{blue} Monetary} = \begin{bmatrix}
0.413 \\ 
0.582  \\
-0.007  \\
0.247  \\
0.216  \\
-0.718  \\
0.147  \\
0.051 \\
\end{bmatrix} 
\end{align*}

**Note**: 
* Word vectors are also called **(word) embeddings** or **(neural) word representations**

* They are a **distributed** representation

* We can do arithmetic with word vectors!

<div align="center">
<img src="images/WordEmbeddings.jpg" alt="Word Embeddings" width="450"/>
</div>

### 2. Word2vec
---
#### Word2vec: A Model for Learning Word Vectors

* **Word2vec** is a framework for learning word vectors from large text corpora ([Mikolov et al.,2013](https://arxiv.org/pdf/1301.3781))

* **Idea**: 
  * A large corpus of text: a long list of words 

  * Every word in a fixed vocabulary is represented by a **vector**
 
  * Go through each position t in the text, which has a *center word* $c$ and *context (“outside”) words* $o$

  * Use the similarity of the word vectors for $c$ and $o$ to calculate the probability of $o$ given $c$ (or vice versa)

  * Keep **adjusting** the word vectors to **maximize this probability** 
<div align="center">
<img src="images/word2vec_model.png" alt="Word2vec" width="400"/>
</div>

* For each position $t=1,2,\cdots, T$, given center word $c_t$, predicting the context words $o_{t+j}$ within a window of size $m$ (e.g., $m=2$ means two words to the left and two words to the right), Word2vec maximizes the following **objective function**:
  $$J(\theta) = -\sum_{t=1}^{T} \sum_{-m\leq j \leq m, j \neq 0} \log P(o_{t+j} | c_t; \theta)$$ 


### Word Vectors using `gensim` in Python 
---
* `Gensim` 是一个流行的 Python 自然语言处理库，专注于主题建模和文档相似度等任务。

* `KeyedVectors` 是 `gensim` 用于高效存储和操作词向量（如 `Word2Vec`、`FastText`、`GloVe` 等预训练词向量）的核心类

##### 中文词向量模型
* 中文文本词向量：[`Text2vec`](https://github.com/shibing624/text2vec)
  * [腾讯中文词向量`word2vec-tencent-chinese`](https://modelscope.cn/models/lili666/text2vec-word2vec-tencent-chinese)
```python
    pip install text2vec
```

In [ ]:
# Word Vector Examples 
# pip install gensim opencc-python-reimplemented 
from gensim.models.word2vec import Word2Vec
from gensim.models import KeyedVectors
import os
import gensim.downloader as api
import zipfile
import urllib.request
import jieba  # 中文分词工具
import numpy as np 

In [58]:
#download English and Chinese word vectors 
# --- Load English word vectors ---
#os.environ['GENSIM_DATA_DIR'] = '/Users/happyhome/Research/NLP-models/gensim_data' 
print("Loading English word vectors (FastText)...")
corpus = api.load("text8")
model = Word2Vec(corpus, vector_size=100)

Loading English word vectors (FastText)...


In [59]:
model.wv.most_similar("economics") 

[('keynesian', 0.8104205131530762),
 ('macroeconomics', 0.767795205116272),
 ('sociology', 0.7389475703239441),
 ('neoclassical', 0.7227122187614441),
 ('liberalism', 0.7025279402732849),
 ('hayek', 0.7014248967170715),
 ('anthropology', 0.6971229910850525),
 ('psychology', 0.6930307745933533),
 ('capitalism', 0.6858665347099304),
 ('ethics', 0.6819972395896912)]

In [60]:
# print the local path of the downloaded model
print(model.wv.most_similar(positive=['king', 'woman'], negative=['man'], topn=3))
print(model.wv.most_similar(positive=['beijing', 'china'], negative=['history'], topn=3))
print(model.wv.most_similar(positive=['summer', 'china'], negative = ['barbecue'], topn=3))

[('queen', 0.6572537422180176), ('prince', 0.6123291850090027), ('empress', 0.6052742600440979)]
[('shanghai', 0.6718805432319641), ('guangzhou', 0.6644915342330933), ('taiwan', 0.6498435139656067)]
[('japan', 0.5749789476394653), ('decline', 0.5173664093017578), ('europe', 0.5158505439758301)]


In [61]:
from text2vec import Word2Vec

In [62]:
# --- Load Chinese word vectors (Tencent AI Lab) ---
def compute_emb(model):
    # Embed a list of sentences
    sentences = [
        '卡',
        '银行卡',
    ]
    sentence_embeddings = model.encode(sentences, show_progress_bar=True, normalize_embeddings=True)
    print(type(sentence_embeddings), sentence_embeddings.shape)

    # The result is a list of sentence embeddings as numpy arrays
    for sentence, embedding in zip(sentences, sentence_embeddings):
        print("Sentence:", sentence)
        print("Embedding shape:", embedding.shape)
        print("Embedding head:", embedding[:10])
        print()

if __name__ == "__main__":
    # 中文词向量模型(word2vec)，中文字面匹配任务和冷启动适用
    w2v_model = Word2Vec("w2v-light-tencent-chinese")
    compute_emb(w2v_model)

2025-06-07 10:08:59.073 | INFO     | text2vec.word2vec:__init__:84 - Load pretrained model:w2v-light-tencent-chinese, path:/Users/happyhome/.text2vec/datasets/light_Tencent_AILab_ChineseEmbedding.bin
2025-06-07 10:08:59.562 | DEBUG    | text2vec.word2vec:__init__:97 - Load w2v from /Users/happyhome/.text2vec/datasets/light_Tencent_AILab_ChineseEmbedding.bin, spend 0.49 sec
Word2Vec Embeddings: 100%|██████████| 2/2 [00:00<00:00, 89.47it/s]

<class 'numpy.ndarray'> (2, 200)
Sentence: 卡
Embedding shape: (200,)
Embedding head: [ 0.06761453 -0.10960816 -0.04829824  0.0156597  -0.09412017 -0.04805465
 -0.03369278 -0.07476041 -0.01600934  0.03106228]

Sentence: 银行卡
Embedding shape: (200,)
Embedding head: [ 0.01032454 -0.13564903 -0.00089282  0.02286329 -0.03501284  0.00987683
  0.02884413 -0.03491557  0.02036332  0.04516884]



In [74]:
model_path = "/Users/happyhome/.text2vec/datasets/light_Tencent_AILab_ChineseEmbedding.bin"
model_chn = KeyedVectors.load_word2vec_format(model_path, binary=True)
print(model_chn.most_similar("中国", topn = 10))

[('其他国家', 0.7651227712631226), ('国家', 0.7572987675666809), ('美国', 0.7531552910804749), ('我国', 0.7485135793685913), ('日本', 0.7391061186790466), ('国内', 0.7384742498397827), ('大国', 0.7368020415306091), ('中国政府', 0.7312999963760376), ('印度', 0.7231327295303345), ('外国', 0.7215994000434875)]


In [ ]:
print(model_chn.most_similar(positive=['北京', '中国'], negative=['历史'], topn=3))

In [ ]:
print(model_chn.most_similar(positive=['中国人民大学', '北京理工大学'], negative=['北京大学'], topn=5))

In [ ]:
print(model_chn.most_similar(positive=['中国人民大学', '清华大学'], negative=['北京理工大学'], topn=5))

In [ ]:
print(model_chn.most_similar(positive=['苹果公司', '索尼'], negative=['日本'], topn=5))

In [ ]:
print(model_chn.most_similar(positive=['华为公司', '中国'], negative=['美国'], topn=5))

In [ ]:
print(model_chn.most_similar(positive=['自由', '健身房'], negative=['工地'], topn=5))

### 3. GloVe （Global Vectors for Word Representation）
---

* **GloVe** is another framework for learning word vectors ([Pennington et al., 2014](https://nlp.stanford.edu/pubs/glove.pdf)) [[Project page]](https://nlp.stanford.edu/projects/glove/)
  
* **Idea**: 
  * A large corpus of text: a long list of words 

  * Every word in a fixed vocabulary is represented by a **vector**

  * Use the **co-occurrence matrix** to calculate the probability of $o$ given $c$ (or vice versa)
    * The `Word2Vec` is a *predictive model* that learns word vectors by predicting context words given a center word.
    * The `GloVe` is a *count-based model* that learns word vectors by factorizing the co-occurrence matrix of words.

* **Optimization**:
  * For word vectors $w_i$ and context vectors $w_j$, GloVe uses the **dot product** to approximate the log of number of co-occurrences of $w_i$ and $w_j$: $$w_i^T w_j + b_i + \tilde{b}_j \approx \log(X_{ij})$$ 

  * It captures the **semantic relationships** using **ratios of probabilities**, not just raw counts. 

  * **Objective function**: $$J = \sum_{i=1}^{V} \sum_{j=1}^{V} f(X_{ij}) (w_i^T w_j + b_i + \tilde{b}_j - \log(X_{ij}))^2$$
    * where $f(X_{ij})$ is a weighting function that down-weights the influence of very frequent co-occurrences.

#### GloVe模型应用于中文文本的限制
##### **GloVe** 的核心是基于全局词共现统计构建词向量。其训练流程为：

* 首先构建词共现矩阵（Word Co-Occurrence Matrix），统计语料中词语在固定窗口内的共现次数；
  
* 然后通过矩阵分解（如最小二乘法）将高维共现矩阵映射到低维向量空间，从而捕捉词语间的语义关联。

> 这一过程高度依赖精准的词语边界划分，因为共现统计的最小单位是 “词”。
##### 中文分词的天然挑战

* **中文是无空格分隔的语言，词语由连续字符组成**，需通过分词（Word Segmentation）确定词边界。然而：
分词标准不统一：不同分词工具（如 Jieba、THULAC）对同一文本可能产出不同结果（例如 “北京大学” 可能被切分为 “北京 / 大学” 或整体作为一词），导致共现矩阵统计偏差。

* **未登录词问题**：新词（如网络流行语 “躺平”）或专业术语难以被传统分词模型覆盖，易被错误切分或忽略，影响共现矩阵的完整性。

* **分词误差的传导效应**：分词错误会直接导致共现矩阵中的统计错误，进而影响词向量的准确性。例如，若 “云计算” 被错误切分为 “云 / 计算”，则 “云” 与 “计算” 的共现频率会被高估，而 “云计算” 作为独立词的语义可能无法被正确捕捉。

> 相比之下，英文等拼音语言天然以空格分隔词语，预处理步骤更简单，共现统计更直接，因此 GloVe 在英文中应用更广泛。

##### 替代方案

* **字符级建模（Character-Level）**:
  * **规避分词依赖：** 字符是中文的最小语义单位（如 “山”“水” 本身具有独立含义），无需分词即可直接处理，避免了分词误差。
  * **泛化能力强：** 字符集相对固定（常用汉字约 3000-5000 个），可有效处理未登录词（如人名、地名、新词），只需将其拆解为字符组合即可。
  * **计算效率高：** 字符级模型的词表规模通常小于词语级（中文词语数量可达百万级别），训练速度更快

* **预训练模型**:
  * `Word2Vec` 的 `Skip-gram/CBOW` 模型可直接基于字符训练，无需分词。如 “人工智能” 可被拆分为 “人”“工”“智”“能” 四个字符，模型通过上下文字符预测目标字符，间接捕捉词语语义。
  * [`FastText`](https://fasttext.cc/) 进一步引入 “字符 n-gram”（如 “人工”“工智”“智能”），在字符级基础上融合局部词语结构信息，兼顾泛化性与语义准确性。

In [77]:
from gensim.scripts.glove2word2vec import glove2word2vec
from gensim.models import KeyedVectors

# Convert glove format to word2vec format
glove_directory = '/Users/happyhome/GloVe_Models/glove.6B/'
glove_input_file = glove_directory+'glove.6B.100d.txt'
word2vec_output_file = glove_directory+'glove.6B.100d.word2vec.txt'

# Only needs to run once
glove2word2vec(glove_input_file, word2vec_output_file)

# Load converted model
model = KeyedVectors.load_word2vec_format(word2vec_output_file, binary=False)


/var/folders/l9/dnh5jmfj53q02s7qff8c36r40000gn/T/ipykernel_19585/1712380481.py:10: DeprecationWarning: Call to deprecated `glove2word2vec` (KeyedVectors.load_word2vec_format(.., binary=False, no_header=True) loads GLoVE text vectors.).
  glove2word2vec(glove_input_file, word2vec_output_file)


In [79]:
# Most similar words
print("Most similar to 'king':")
print(model.most_similar('king', topn=5))

# Vector arithmetic: king - man + woman ≈ queen
print("\nking - man + woman ≈ ?")
result = model.most_similar(positive=['king', 'woman'], negative=['man'], topn=1)
print(result)

# Cosine similarity
similarity = model.similarity('china', 'france')
print(f"\nSimilarity between 'china' and 'france': {similarity:.4f}")


Most similar to 'king':
[('prince', 0.7682329416275024), ('queen', 0.7507689595222473), ('son', 0.7020888328552246), ('brother', 0.6985775828361511), ('monarch', 0.6977890729904175)]

king - man + woman ≈ ?
[('queen', 0.7698541283607483)]

Similarity between 'china' and 'france': 0.4527


In [88]:
# Load pre-trained fastText Chinese vectors (can take several minutes and ~8GB RAM)
print("Loading fastText vectors...")
model = KeyedVectors.load_word2vec_format("/Users/happyhome/fast_text_zh/cc.zh.300.vec", binary=False)
print("Model loaded.")

Loading fastText vectors...
Model loaded.


In [90]:
# Function: segment and clean Chinese sentence
def tokenize_chinese(text):
    return list(jieba.cut(text))

# Function: average sentence embedding
def sentence_vector(sentence, model):
    words = tokenize_chinese(sentence)
    vecs = [model[word] for word in words if word in model]
    if vecs:
        return sum(vecs) / len(vecs)
    else:
        return None
# Example 1: nearest words
word = "银行"
if word in model:
    print(f"Words similar to {word}:")
    for w, score in model.most_similar(word, topn=5):
        print(f"{w}: {score:.4f}")
# Example 2: word analogy: 王 - 男 + 女 ≈ ?
print("\nAnalogy: 王 - 男 + 女 ≈ ?")
if all(w in model for w in ['王', '男', '女']):
    result = model.most_similar(positive=['王', '女'], negative=['男'], topn=1)
    print(result)
# Example 3: sentence similarity
s1 = "我喜欢人工智能"
s2 = "机器学习是人工智能的核心"
v1 = sentence_vector(s1, model)
v2 = sentence_vector(s2, model)
if v1 is not None and v2 is not None:
    cosine_sim = v1.dot(v2) / (np.linalg.norm(v1) * np.linalg.norm(v2))
    print(f"\nSimilarity between:\n  \"{s1}\"\n  \"{s2}\" = {cosine_sim:.4f}")
else:
    print("Sentence vector failed.")

Words similar to 银行:
银行业: 0.7677
农行: 0.7499
行长: 0.7102
总行: 0.7098
浦发: 0.6945

Analogy: 王 - 男 + 女 ≈ ?
[('万凰', 0.5156221389770508)]

Similarity between:
  "我喜欢人工智能"
  "机器学习是人工智能的核心" = 0.5196


## 4.  RNNs (Recurrent Neural Networks, 循环神经网络)

---

#### Why Recurrent Neural Networks (RNNs)?

- NLP deals with **sequential data** (words, sentences).
  
- Traditional feed=forward neural networks can't handle **variable-length input** or remember **context**.
  
- RNNs maintain a **hidden state** that evolves over time, making them ideal for sequence modeling.



#### What is an RNN?
- **Core Idea**: Apply the same *weights* $W$ repeatedly across time steps to process sequences.
- 
- At each time step $ t $, the RNN takes input $ x_t $ and the previous **hidden state** $ h_{t-1} $, and computes:

$$
h_t = \tanh(W_{hx}x_t + W_{hh}h_{t-1} + b_h)
$$

- The hidden state captures past information.

<div align="center">
<img src="images/RNNs.webp" alt="RNNs" width="500"/>
</div>


#### Training an RNN Language Model 

<div align="center">
<img src="images/RNNs_training.png" alt="RNNs Training" width="500"/>
</div>

* $J^{(i)}(\theta)$ is the negative log probability of the next word given the previous words in the sequence. 

#### Applications of RNNs 

> **You can train an RNN Language Model on any kind of text, then generate new text in that style!**


## LSTM: Long Short-Term Memory (长短期记忆网络)
---

- **Limitations of Vanilla RNNs**：
  - **Vanishing gradients** make it hard to learn long-term dependencies.
  - Struggles with remembering information over many time steps.

#### [LSTM](https://medium.com/@ottaviocalzone/an-intuitive-explanation-of-lstm-a035eb6ab42c) architecture 

<div align="center">
<img src="images/LSTM_structure.webp" alt="LSTM Architecture" width="500"/>
</div>

**LSTM** introduces **gates** to control the flow of information:
- **Forget gate** $ f_t $: Decides **what to forget** from the cell state. $$f_t = \sigma(W_f \cdot [h_{t-1}, x_t] + b_f)$$

  - **Output**: values between 0 (forget) and 1 (keep).
  - Multiplies the old cell state $ c_{t-1}$ to **discard irrelevant info**.
- **Input gate** $ i_t $ + **Candidate** $\tilde{c}_t$
Decides **what new information to store**. $$i_t = \sigma(W_i \cdot [h_{t-1}, x_t] + b_i)$$ $$\tilde{c}_t = \tanh(W_c \cdot [h_{t-1}, x_t] + b_c)$$
  - $i_t $: how much of $\tilde{c}_t$ to add to the cell state. 
- **Output gate** $ o_t $ Decides **what to output** at this step. $$o_t = \sigma(W_o \cdot [h_{t-1}, x_t] + b_o)$$ $$h_t = o_t \odot \tanh(c_t)$$

  - $h_t$: the **hidden state**, used in the next step or output.
- Maintains a **cell state** $ c_t $ to store long-term memory and update it: $$c_t = f_t \odot c_{t-1} + i_t \odot \tilde{c}_t$$
  - Combines **old memory** and **new memory** based on gates.
  

#### ELMo: Embeddings from Language Models
---

- **ELMo** (embeddings from language model) is a word embedding method for representing a sequence of words as a corresponding sequence of vectors. [[Paper]](https://arxiv.org/abs/1802.05365)[[Wiki]](https://en.wikipedia.org/wiki/ELMo?utm_source=chatgpt.com)

- It is a bidirectional LSTM which takes character-level as inputs and produces word-level embeddings, trained on a corpus of about `30 million` sentences and `1 billion words`.

- It was an instrumental step in the evolution towards transformer-based language modelling.


## 5. Attention Mechanisms and Transformers
---
#### What is Attention?
- **Attention** is a mechanism that allows models to focus on specific parts of the input sequence when making predictions.

>**Core Idea**: on each step of the decoder, use **direct connection to the encoder** to **focus on a particular part** of the source sequence

<div align="center">
<img src="images/sequence_to_sequence_attention1.png" alt="Attention Mechanism" width="450"/>
<img src="images/sequence_to_sequence_attention2.png" alt="Attention Mechanism" width="450"/>
<img src="images/sequence_to_sequence_attention3.png" alt="Attention Mechanism" width="450"/>
<img src="images/sequence_to_sequence_attention_final.png" alt="Attention Mechanism" width="450"/>
</div>

#### Attention Mechanism in Equations

- We have encoder hidden states $h_1, \ldots, h_N \in \mathbb{R}^h$

- On timestep $t$, we have decoder hidden state $s_t \in \mathbb{R}^h$

- We get the attention scores $e^t$ for this step:

  $$
  e^t = [s_t^\top h_1, \ldots, s_t^\top h_N] \in \mathbb{R}^N
  $$

- We take softmax to get the attention distribution $\alpha^t$ for this step (this is a probability distribution and sums to 1):

  $$
  \alpha^t = \text{softmax}(e^t) \in \mathbb{R}^N
  $$

- We use $\alpha^t$ to take a weighted sum of the encoder hidden states to get the attention output $a_t$:

  $$
  a_t = \sum_{i=1}^N \alpha^t_i h_i \in \mathbb{R}^h
  $$

- Finally we concatenate the attention output $a_t$ with the decoder hidden state $s_t$ and proceed as in the non-attention seq2seq model:

  $$
  [a_t; s_t] \in \mathbb{R}^{2h}
  $$


#### Attention is a Transformative Idea!

-  **Improves NMT (Neural Machine Translation)**: Major boost in neural machine translation performance  (Google Translate, 2014)
  
- **Dynamic Focus**: Decoder attends to relevant parts of the input at each step  
  
- **Human-like**: Mimics how humans refer back during translation  
  
- **Solves Bottleneck**: Avoids compressing all source info into one fixed vector  
-  **Mitigates Vanishing Gradient**: Provides shortcut to distant tokens  
-  **Interpretability**: Attention weights show what the model is "looking at" 
   
- **Free Alignment**: Learns soft word alignment without supervision  
  - The network learns to align words in the source and target languages without explicit alignment data! (unsupervised learning)


#### **Attention** is a *general* Deep Learning Technique 

> **General Definition of Attention**: Given a set of vector **values**, and vector of **query**, **attention** is a technique to compute a weighted sum of the values, dependent on the query. 

#### Why Transformers?
- **Limitations of RNNs**:
  - Sequential processing makes them slow. Future RNN hidden states can't be computed in full before past RNN hidden states are computed.
  
  - Hard to parallelize training.
  
  - Hard to capture long-range dependencies.
- **Transformers** use **self-attention** to process all words in parallel, capturing relationships regardless of distance. [[Attention is All You Need!]](https://arxiv.org/abs/1706.03762) [[Google Blog]](https://research.google/blog/transformer-a-novel-neural-network-architecture-for-language-understanding/)

- **Transformers** even show promise in tasks other than NLP, such as image processing. 




### Self-Attention: Queries, Keys, and Values from the Same Input 
---
<div align="center">
<b>Contextualized Embeddings</b><br>
<img alt="Contextualized embeddings" caption="Diagram showing how self-attention updates raw token embeddings (upper) into contextualized embeddings (lower) to create representations that incorporate information from the whole sequence" src="images/chapter03_contextualized-embedding.png" id="contextualized-embeddings", width=380/>
</div>

Let $\boldsymbol{W}\in \mathbb{R}^{n\times|V|}$ be a sequence of words in vocabulary $V$, e.g., *economics needs revolution*. 

For each row $\boldsymbol{W}_i \in \mathbb{R}^{|V|}$ (a row of $\boldsymbol{W}$), $\boldsymbol{x}_i = E_{\boldsymbol{W}_i}$, $E\in \mathbb{R}^{d\times |V|}$ is an embedding matrix

1. Transform each word embedding with weight matrices $Q,\,K,\,V \in \mathbb{R}^{d\times d}$ $${\color{red}\boldsymbol{q}_i} = Q\boldsymbol{x}_i \text{\color{red}(queries)}\quad\quad {\color{blue}\boldsymbol{k}_i} = K\boldsymbol{x}_i \text{\color{blue}(keys)}\quad\quad {\color{green}\boldsymbol{v}_i} = V\boldsymbol{x}_i \text{\color{green}(values)}$$

2. Compute the pairwise similarities between keys and queries; normalized with `softmax` to get attention weights:
   $$ \boldsymbol{e}_{ij} = {\color{red}\boldsymbol{q}_i}^\top {\color{blue}\boldsymbol{k}_j} \quad\quad \alpha_{ij} = \frac{\exp(\boldsymbol{e}_{ij})}{\sum_{k=1}^{n}\exp(\boldsymbol{e}_{ik})}$$

3. Compute output for each word as a weighted sum of **values**: $$\boldsymbol{o}_i = \sum_{j=1}^{n} \alpha_{ij} {\color{green}\boldsymbol{v}_j}$$

#### The Transformer Architecture [(Vaswani et al., 2017)](https://arxiv.org/pdf/1706.03762)
<div align="center">
<img src="images/attention_paper_impact.png" alt="Transformer Architecture Overview" width="350"/>
</div>


<div align="center">

<img src="images/transformer_architecture.png" alt="Transformer Architecture" height="400"/>
</div>

##### Let go over the key components of the Transformer architecture one by one!

### An Anatomy of the Transformer Architecture
---

**Reference**: 

* [NLP with Transformers: Chapter 03-Transformer Anatomy](https://github.com/nlp-with-transformers/notebooks)
* [Transformers for NLP: Chapter 02-Multi-Head Attention]([Chapter02/Multi_Head_Attention_Sub_Layer.ipynb](https://github.com/Denis2054/Transformers-for-NLP-2nd-Edition/blob/main/Chapter02/Multi_Head_Attention_Sub_Layer.ipynb))

<div align="center">
<img alt="transformer-encoder-decoder" caption="Encoder-decoder architecture of the transformer, with the encoder shown in the upper half of the figure and the decoder in the lower half" src="images/chapter03_transformer-encoder-decoder.png" id="transformer-encoder-decoder", width=500/>
</div>


#### The Encoder 
> **Encoder Layer** receives sequence of embeddings and feeds them through sub-layers:
> * **Multi-Head Self-Attention** layer
> * **Feed-Forward Neural Network** layer

<div align="center">
<img alt="encoder-zoom" caption="Zooming into the encoder layer" src="images/chapter03_encoder-zoom.png" id="encoder-zoom", width=400/>
</div>

##### Self-Attention Layer 
> **Self-Attention** layer: Use the whole sequence to compute a *weighted average* of each embedding. 
 1.  Project each token embedding into three vectors: **query** ($\mathbf{Q}$), **key** ($\mathbf{K}$), and **value** ($\mathbf{V}$).
  
 2. Compute **attention scores** and **attention weights** using `softmax` function.$$\text{Attention}(\mathbf{Q}, \mathbf{K}, \mathbf{V}) = \text{softmax}\left(\frac{\mathbf{Q} \cdot \mathbf{K}^\top}{\sqrt{d_k}}\right) \cdot \mathbf{V}$$
      - $\sqrt{d_k}$ is a scaling factor to prevent large dot products.

 3. Update the token embeddings: Compute the output as a **weighted sum** of the value vectors, using the attention weights. 

<div align="center">

<img alt="Operations in scaled dot-product attention" height="100" caption="Operations in scaled dot-product attention" src="images/chapter03_attention-ops.png" id="attention-ops"/>
</div>

Let's visualize this process using `BertViz`.

In [2]:
##bertviz environment setup 
!pip install bertviz transformers torch --quiet

In [ ]:
#hide_output
from transformers import AutoTokenizer
from bertviz.transformers_neuron_view import BertModel
from bertviz.neuron_view import show
model_ckpt = "bert-base-uncased" #英文基础版本模型，所有英文单词小写
tokenizer = AutoTokenizer.from_pretrained(model_ckpt)
model = BertModel.from_pretrained(model_ckpt)
text = "time flies like an arrow"

In [7]:
show(model, "bert", tokenizer, text, display_mode="light", layer=0, head=8)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
#show tokens for the text 
tokens = tokenizer.tokenize(text)
print("Tokens:", tokens)
#show word embeddings 
input_ids = tokenizer.convert_tokens_to_ids(tokens)
embeddings = model.embeddings.word_embeddings.weight[input_ids]
print("Embeddings shape:", embeddings.shape)

Tokens: ['time', 'flies', 'like', 'an', 'arrow']
Embeddings shape: torch.Size([5, 768])


In [ ]:
# Self-attention operations 
#display the input ids
inputs = tokenizer(text, return_tensors="pt", add_special_tokens=False)
print(inputs.input_ids)
inputs1 = tokenizer(text, return_tensors="pt", add_special_tokens=True) #with [CLS] and [SEP] tokens
print(inputs1.input_ids)

tensor([[ 2051, 10029,  2066,  2019,  8612]])
tensor([[  101,  2051, 10029,  2066,  2019,  8612,   102]])


In [11]:
from torch import nn
from transformers import AutoConfig

config = AutoConfig.from_pretrained(model_ckpt)
token_emb = nn.Embedding(config.vocab_size, config.hidden_size) #vocab_size and embedding size
token_emb

Embedding(30522, 768)

In [12]:
# input IDs to embeddings
inputs_embeds = token_emb(inputs.input_ids)
inputs_embeds.size()

torch.Size([1, 5, 768])

In [27]:
import torch 
from math import sqrt 
import torch.nn.functional as F
def scaled_dot_product_attention(query, key, value):
    dim_k = query.size(-1)
    scores = torch.bmm(query, key.transpose(1, 2)) / sqrt(dim_k)
    weights = F.softmax(scores, dim=-1)
    return torch.bmm(weights, value)

In [24]:
query = key = value = inputs_embeds 
dim_k = key.size(-1) #size of embeddings
print(dim_k)
print(query.shape)
print(key.transpose(1, 2).shape)

768
torch.Size([1, 5, 768])
torch.Size([1, 768, 5])


In [30]:
scores = torch.bmm(query, key.transpose(1, 2)) / sqrt(dim_k)
print(scores)

tensor([[[26.5318, -1.1530,  1.4374, -0.5730,  0.3117],
         [-1.1530, 27.8085, -1.7578,  0.7391, -0.1518],
         [ 1.4374, -1.7578, 27.7157,  1.9828,  0.1467],
         [-0.5730,  0.7391,  1.9828, 28.7732,  0.5681],
         [ 0.3117, -0.1518,  0.1467,  0.5681, 30.9588]]],
       grad_fn=<DivBackward0>)


In [ ]:
#understand the dimensions of using softmax 
row = scores[0][0,:]
row = row.view(1, 1, 5)
#apply softmax to get attention weights
print(F.softmax(row, dim=-1))
print(F.softmax(row, dim=1))

tensor([[[1.0000e+00, 9.4757e-13, 1.2636e-11, 1.6926e-12, 4.0995e-12]]],
       grad_fn=<SoftmaxBackward0>)
tensor([[[1., 1., 1., 1., 1.]]], grad_fn=<SoftmaxBackward0>)


In [64]:
weights = F.softmax(scores, dim=-1) #
print(weights)
print(weights.sum(dim=-1)) #sum of weights for each query vector

tensor([[[1.0000e+00, 9.4757e-13, 1.2636e-11, 1.6926e-12, 4.0995e-12],
         [2.6434e-13, 1.0000e+00, 1.4438e-13, 1.7535e-12, 7.1943e-13],
         [3.8681e-12, 1.5843e-13, 1.0000e+00, 6.6734e-12, 1.0641e-12],
         [1.7994e-13, 6.6826e-13, 2.3178e-12, 1.0000e+00, 5.6321e-13],
         [4.8990e-14, 3.0819e-14, 4.1542e-14, 6.3308e-14, 1.0000e+00]]],
       grad_fn=<SoftmaxBackward0>)
tensor([[1., 1., 1., 1., 1.]], grad_fn=<SumBackward1>)


In [65]:
#attention output (back to embeddings)
attn_outputs = torch.bmm(weights, value)
print(attn_outputs)  # shape should be (batch_size, seq_len, embedding_dim)
attn_outputs.shape

tensor([[[-1.7070,  1.5176, -0.9279,  ..., -0.4497, -1.2688,  1.1617],
         [ 0.1954, -0.1177, -0.0234,  ..., -0.8323,  0.4386,  1.3718],
         [ 1.5429, -0.0116,  0.2128,  ...,  1.6273, -0.2630,  0.4664],
         [ 0.3452, -0.2181,  1.3202,  ..., -0.2276,  0.7309, -0.7244],
         [-0.0428,  1.1080, -0.6644,  ...,  1.3628, -0.6486, -0.5980]]],
       grad_fn=<BmmBackward0>)


torch.Size([1, 5, 768])

In [66]:
# use the define function 
attn_outputs = scaled_dot_product_attention(query, key, value)
print(attn_outputs.shape)  # should be (batch_size, seq_len, embedding_dim)

torch.Size([1, 5, 768])


#### Multi-Head Attention

> **Multi-Head Attention**: Instead of a single attention mechanism, we use multiple attention heads to capture different aspects of the input. Each head has its own set of *independent* **query**, **key**, and **value** projections. 

* Why multi-head attention?
  - **Diversity**: Different heads can focus on different parts of the input.
  
  - **Parallelization**: Heads can be computed in parallel, improving efficiency.

<div align="center">
<img alt="Multi-head attention" height="150" caption="Multi-head attention" src="images/chapter03_multihead-attention.png" id="multihead-attention"/>
</div>

$$
\text{Head's Dimension} = \frac{\text{Embedding Dimension}}{\text{Number of Heads}}
$$

In [73]:
#define the attention head and multi-head attention
class AttentionHead(nn.Module):
    def __init__(self, embed_dim, head_dim):
        super().__init__()
        self.q = nn.Linear(embed_dim, head_dim) 
        self.k = nn.Linear(embed_dim, head_dim)
        self.v = nn.Linear(embed_dim, head_dim)

    def forward(self, hidden_state):
        attn_outputs = scaled_dot_product_attention(
            self.q(hidden_state), self.k(hidden_state), self.v(hidden_state))
        return attn_outputs
    
class MultiHeadAttention(nn.Module):
    def __init__(self, config):
        super().__init__()
        embed_dim = config.hidden_size
        num_heads = config.num_attention_heads
        head_dim = embed_dim // num_heads #head_dimension is the ratio of embedding and number of heads
        self.heads = nn.ModuleList(
            [AttentionHead(embed_dim, head_dim) for _ in range(num_heads)]
        )
        self.output_linear = nn.Linear(embed_dim, embed_dim)

    def forward(self, hidden_state):
        x = torch.cat([h(hidden_state) for h in self.heads], dim=-1)
        #print(x)
        # Apply the output linear layer to combine the heads
        x = self.output_linear(x)
        return x

In [75]:
multihead_attn = MultiHeadAttention(config)
attn_output = multihead_attn(inputs_embeds)   
attn_output.size() 

torch.Size([1, 5, 768])

#### The Feed-Forward Layer

* The **Feed-Forward Layer** is a simple two-layer fully connected neural network:
  
  * It processes each embedding independently, often referred as a **position-wise feed-forward layer**. 
  
  * A *rule of thumb* is let the hidden layer size be 4 times the embedding size, i.e., $d_{ff} = 4 \times d_{model}$. 

  * The activation function is usually **GELU** (Gaussian Error Linear Unit). 



In [ ]:
class FeedForward(nn.Module):
    def __init__(self, config):
        super().__init__()
        self.linear_1 = nn.Linear(config.hidden_size, config.intermediate_size)
        self.linear_2 = nn.Linear(config.intermediate_size, config.hidden_size)
        self.gelu = nn.GELU()
        self.dropout = nn.Dropout(config.hidden_dropout_prob) #randomly assigning 0 to some neurons to avoid overfitting
        
    def forward(self, x):
        x = self.linear_1(x)
        x = self.gelu(x)
        x = self.linear_2(x)
        x = self.dropout(x)
        return x

# apply feed-forward network to the attention output
print(config.hidden_size, config.intermediate_size, config.hidden_dropout_prob)
feed_forward = FeedForward(config)
ff_outputs = feed_forward(attn_outputs)
ff_outputs.size()

768 3072 0.1


torch.Size([1, 5, 768])

#### Adding Layer Normalization and Skip Connections

>1. **Layer Normalization**: Normalizes the input in the batch to have **zero mean** and **unit variance**:
    >   - **Post layer normalization**: Originally used in the Transformer paper, it applies normalization in between the skip connections.
    >   - **Pre layer normalization**: A more common approach that places layer within the span of the skip connections. 

>2. **Skip Connections**: pass a tensor to the next layer without processing and add it to the processed tensor.  
<div align="center">
<img alt="Transformer layer normalization" height="250" caption="Different arrangements of layer normalization in a transformer encoder layer" src="images/chapter03_layer-norm.png" id="layer-norm"/>
</div>

In [77]:
# pre-layer normalization
class TransformerEncoderLayer(nn.Module):
    def __init__(self, config):
        super().__init__()
        self.layer_norm_1 = nn.LayerNorm(config.hidden_size)
        self.layer_norm_2 = nn.LayerNorm(config.hidden_size)
        self.attention = MultiHeadAttention(config)
        self.feed_forward = FeedForward(config)

    def forward(self, x):
        # Apply layer normalization and then copy input into query, key, value
        hidden_state = self.layer_norm_1(x)
        # Apply attention with a skip connection
        x = x + self.attention(hidden_state)
        # Apply feed-forward layer with a skip connection
        x = x + self.feed_forward(self.layer_norm_2(x))
        return x
    
encoder_layer = TransformerEncoderLayer(config)
inputs_embeds.shape, encoder_layer(inputs_embeds).size()

(torch.Size([1, 5, 768]), torch.Size([1, 5, 768]))

#### Positional Encoding
The **multi-head attention layer** is effectively a fancy weighted sum, the information on token position is lost. 

> **Positional Encoding**: Since the Transformer has no inherent notion of word order, we add positional encodings to the input embeddings to provide information about the position of each token in the sequence.
$$
PE_{(pos, 2i)} = \sin\left(\frac{pos}{10000^{2i/d}}\right) \quad \text{and} \quad PE_{(pos, 2i+1)} = \cos\left(\frac{pos}{10000^{2i/d}}\right)
$$
where $d$ is the dimension of the embeddings, $pos$ is the position of the token in the sequence, and $i$ is the index of the dimension.

In [78]:
class Embeddings(nn.Module):
    def __init__(self, config):
        super().__init__()
        self.token_embeddings = nn.Embedding(config.vocab_size, 
                                             config.hidden_size)
        self.position_embeddings = nn.Embedding(config.max_position_embeddings,
                                                config.hidden_size)
        self.layer_norm = nn.LayerNorm(config.hidden_size, eps=1e-12)
        self.dropout = nn.Dropout()

    def forward(self, input_ids):
        # Create position IDs for input sequence
        seq_length = input_ids.size(1)
        position_ids = torch.arange(seq_length, dtype=torch.long).unsqueeze(0)
        # Create token and position embeddings
        token_embeddings = self.token_embeddings(input_ids)
        position_embeddings = self.position_embeddings(position_ids)
        # Combine token and position embeddings
        embeddings = token_embeddings + position_embeddings
        embeddings = self.layer_norm(embeddings)
        embeddings = self.dropout(embeddings)
        return embeddings
    
embedding_layer = Embeddings(config)
embedding_layer(inputs.input_ids).size()

torch.Size([1, 5, 768])

In [79]:
#putting all these together to construct the encoding layer
class TransformerEncoder(nn.Module):
    def __init__(self, config):
        super().__init__()
        self.embeddings = Embeddings(config)
        self.layers = nn.ModuleList([TransformerEncoderLayer(config) 
                                     for _ in range(config.num_hidden_layers)])

    def forward(self, x):
        x = self.embeddings(x)
        for layer in self.layers:
            x = layer(x)
        return x
    
encoder = TransformerEncoder(config)
encoder(inputs.input_ids).size()

torch.Size([1, 5, 768])

### The Decoder 

The main difference between the encoder and the decoder are that decoders have two sub-layers:
* **Masked multi-head self-attention layer**:  Ensures that the tokens we generate at each timestep are only based on the *past outputs and the current token being predicted*. 

* **Encoder-decoder attention layer**:  Performs multi-head attention over the output key and value vectors of the encoder stack, with the *intermediate representations of the decoder acting as the queries*.

<div align="center">
<img alt="Transformer decoder zoom" caption="Zooming into the transformer decoder layer" width=500 src="images/chapter03_decoder-zoom.png" id="decoder-zoom"/> 
</div>

In [80]:
seq_len = inputs.input_ids.size(-1)
mask = torch.tril(torch.ones(seq_len, seq_len)).unsqueeze(0)
print(mask[0])
print(scores.masked_fill(mask == 0, float('-inf')))

tensor([[1., 0., 0., 0., 0.],
        [1., 1., 0., 0., 0.],
        [1., 1., 1., 0., 0.],
        [1., 1., 1., 1., 0.],
        [1., 1., 1., 1., 1.]])
tensor([[[26.5318,    -inf,    -inf,    -inf,    -inf],
         [-1.1530, 27.8085,    -inf,    -inf,    -inf],
         [ 1.4374, -1.7578, 27.7157,    -inf,    -inf],
         [-0.5730,  0.7391,  1.9828, 28.7732,    -inf],
         [ 0.3117, -0.1518,  0.1467,  0.5681, 30.9588]]],
       grad_fn=<MaskedFillBackward0>)


In [ ]:
#adding the mask to the scaled dot-product attention function 
def scaled_dot_product_attention(query, key, value, mask=None):
    dim_k = query.size(-1)
    scores = torch.bmm(query, key.transpose(1, 2)) / sqrt(dim_k)
    if mask is not None:
        scores = scores.masked_fill(mask == 0, float('-inf'))
    weights = F.softmax(scores, dim=-1)
    return torch.bmm(weights, value)

A Recap of the Encoder-Decoder Attention 
---

> ##### Imagine you (`the decoder`) are in class taking an exam. Your task is to predict the next word based on the previous words (`decoder inputs`), which sounds simple but is incredibly hard (try it yourself and predict the next words in a passage of this book).  Fortunately, your neighbor (`the encoder`) has the full text. Unfortunately, they’re a foreign exchange student and the text is in their mother tongue. Cunning students that you are, you figure out a way to cheat anyway. You draw a little cartoon illustrating the text you already have (`the query`) and give it to your neighbor. They try to figure out which passage matches that description (`the key`), draw a cartoon describing the word following that passage (`the value`), and pass that back to you. With this  system in place, you ace the exam.

### 7. BERT and GPTs
---
<div align="center">
<b>Transformer Family Tree</b><br>
<img alt="Transformer family tree" caption="An overview of some of the most prominent transformer architectures" width = 400 src="images/chapter03_transformers-compact.png" id="family-tree"/>

</div>


#### The Encoder Branch: BERT and Other Variants
---
* **BERT** (Bidirectional Encoder Representations from Transformers) is a transformer-based model that uses the encoder part of the transformer architecture
  * It is trained with two objectives of predicting masked tokens in texts: **Masked Language Modeling (MLM)** and **Next Sentence Prediction (NSP)**.
  
  * Dominate NLP tasks such as text classification, named entity recognition, and question answering.

* **BERT Variants**: 
  * **RoBERTa**: A robustly optimized BERT pretraining approach that removes the NSP objective and uses larger mini-batches and longer training.
  
  * **ALBERT**: A lite version of BERT that reduces the number of parameters by sharing weights across layers and factorizing the embedding matrix.
  
  * **DistilBERT**: A smaller, faster, cheaper version of BERT that retains 97% of its language understanding while being 60% faster.
  ......

#### The Decoder Branch: GPTs
---
* **GPT/GPT-2/GPT-3** (Generative Pre-trained Transformer) is a transformer-based model that uses the decoder part of the transformer architecture:
  * It is trained with a **causal language modeling** objective, predicting the next token in a sequence given the previous tokens.

  * GPT-3 has 175 billion parameters. 

  * It is used for text generation, summarization, translation, and other tasks that require generating coherent and contextually relevant text.

#### The Encode-Decoder Branch
---
* **T5** (Text-to-Text Transfer Transformer) 
    * T5 is a transformer model that treats every NLP task as a text-to-text problem, where both input and output are text sequences.
    
* **BART** (Bidirectional and Auto-Regressive Transformers)
    * BART combines BERT and GPT within the encoder-decoder architecture. 

#### Applications of BERT Models
---

* **Text Classification**: Classifying texts into predefined categories, such as sentiment analysis or topic classification.
  * Juhász et al. (2023) fine-tune a BERT model using labeled data on industrial policy texts from GTA (Global Trade Alert) 

<div align="center">
<b>Number of Industrial Policy</b><br>
<img src="images/IP_textcounts.png" alt="IP Text Counts" width="400"/>
</div>

  * [Huang, Wang and Yang (2023)](https://onlinelibrary.wiley.com/doi/epdf/10.1111/1911-3846.12832) develop FinBERT for the finance domain, showing that it outperforms other models in sentiment analysis tasks and identifying discussion topics. 

<div align="center">
<b>FinBERT Model Performance for Sentiment Analysis</b><br>
<img src="images/FinBERT_accuracy.png" alt="FinBERT Sentiment Analysis" width="400"/>
</div>

  * [SsciBERT (2022)](https://link.springer.com/article/10.1007/s11192-022-04602-4) show excellent performance on discipline classification, abstract structure–function recognition, and named entity recognition tasks with the social sciences literature.
  
  * Liu et al. (2024) use fine-tuned BERT-wwm (whole word masking) to map about 100 million online job postings to SOC (Standard Occupational Classification) codes. 
<div align="center">
<img src="images/bert_wwm_job_classification_performance.png" alt="BERT WWM Job Classification Performance" width="500"/>
</div>
  

## LLMs in Social Science Research
---


### 1. Workflow: LLM Applications Across Research Stages

* **Ideation:** LLMs (e.g. GPT) support brainstorming and refining research questions at all stages of idea generation.
  
* **Literature review:** LLMs can suggest keywords, generate search queries, screen abstracts, and summarize articles to accelerate literature discovery.
  
* **Data processing:** For large text data (e.g. interviews, social media), LLMs speed up transcription, translation and labeling.  They can automate qualitative coding and thematic analysis once human codebooks exist.
  
* **Analysis:** LLMs help interpret findings by extracting themes or writing analysis summaries.  Techniques like chain-of-thought prompting improve coding fidelity.  They can also generate data-analysis scripts or calculations on demand.
  
* **Writing & communication:** LLMs assist drafting manuscripts and proposals (e.g. outlines, abstracts, discussions) and improving style.  They suggest phrasing, check grammar, translate text, and even format citations. These tools boost efficiency and writing quality.
  
* **Peer review & publishing:** LLMs could pre-screen submissions, match reviewers by topic, and help reviewers edit language.  However, they cannot yet assess scientific validity.  Studies note LLMs’ strength is in linguistic tasks, while human oversight is needed for substance. (Best practice: use LLMs as *assistants*, not replacements, with clear guidelines.)


**References**:
1. [Haase and Pokutta (2025)](https://arxiv.org/pdf/2506.01839) introduces a structured
framework for understanding the diverse applications of LLM-based agents, ranging from simple data processors
to complex, multi-agent systems capable of simulating emergent social dynamics. 

2. [Novy-Marx and Vlikov (2024)](https://www.nber.org/system/files/working_papers/w33363/w33363.pdf) ("AI-Powered (Finance) Scholarship") illustrates AI’s potential
for enhancing financial research efficiency, but also serves as a cautionary tale, illustrating how it
can be abused to industrialize **HARKing (Hypothesizing After Results are Known)**.
   
3. [Xu et al., (2024)](https://arxiv.org/pdf/2401.11839) surveyed the two-way interaction between AI and Social Science Research. 

4. [Gao et al. (2024)](https://www.nature.com/articles/s41599-024-03611-3)  surveys the landscape of utilizing large language models in agent-based modeling and simulation, discussing their challenges and promising future directions. 

5. [Tsvetkova et al. (2024)](https://www.nature.com/articles/s41562-024-02001-8) review recent research and identify general dynamics and patterns in situations of competition, coordination, cooperation, contagion and collective decision-making in a complex social systems networks of multiple interdependent and interacting humans and intelligent machines constitute complex social systems



### 2. Social Science Tools & Examples

* **Qualitative analysis:** LLMs are used for coding and content analysis of texts (e.g. forum posts, interviews).  For example, ChatGPT was found to reliably suggest inductive coding schemes and assist thematic analysis in health/social forums.  Hybrid workflows keep humans in the loop to refine codebooks while LLMs apply them at scale.
  
  
* **Survey research:** LLMs generate synthetic survey data or “personas” to pilot studies and augment samples.  Researchers use “persona-driven” prompting to simulate demographic response patterns, offering cost-effective data augmentation for attitudes and forecasting.
  
* **Social simulations:** LLMs can simulate virtual populations or agents.  Early work shows GPT-4-based simulations can mimic human experiment responses (e.g. predicting treatment effects across 70 studies).  This enables exploratory “what-if” experiments in sociology, economics or psychology (e.g. modeling group behavior or policy impacts).
  
* **Political science:** LLMs analyze and generate political texts.  They are applied to study rhetoric (speeches, manifestos) and even create campaign materials.  For instance, campaigns already use LLMs to produce AI-generated political ads (e.g. an AI-generated campaign video), which opens new avenues for research on AI-driven political communication.
  
* **Tools & platforms:** Common LLM interfaces include OpenAI’s ChatGPT/GPT-4, Google’s Bard, and Anthropic’s Claude.  Researchers also integrate LLMs via APIs or plugins: e.g. a recent study used GPT-4o via Covidence for screening and used ChatGPT/Scite.ai to clean data, code figures and draft a review.  Emerging tools (custom GPTs, RAG systems) allow linking LLMs to private corpora or databases for context-aware Q\&A.



## 3. Limitations, Ethics & Trends

* **Accuracy & bias:** LLMs can “hallucinate” facts or produce plausible but incorrect content.  Outputs depend heavily on the input prompt and training data.  They inherit biases from their data: if training text under-represents some groups or views, the model’s suggestions will be biased. Their internal reasoning is opaque (“black box”), so results must be carefully validated by researchers.
  
* **Ethical concerns:** Using LLMs raises confidentiality and integrity issues.  Data privacy must be managed (sensitive or unpublished data should not be exposed unnecessarily).  AI-generated text must be vetted for plagiarism and properly attributed.  The community consensus is LLMs **should not be listed as authors**.  Their use must be transparently disclosed, and humans must take responsibility for all scientific content.  Most journals forbid LLMs as co-authors and require any AI assistance to be acknowledged.
  
* **Oversight & guidelines:** Experts stress that LLMs are tools to augment (not replace) expertise.  Human oversight is needed at every stage to check validity.  Proposed best practices include having humans refine prompts, verify outputs against data, and follow ethical guidelines (see examples in peer-review and publication policies).
  
* **Environmental & access issues:** Training and running large LLMs consume huge computational resources, raising sustainability and equity concerns.  Thus, open-source and smaller models (e.g. LLaMA, BLOOM) are emerging to democratize access, and specialized fine-tuning is used to reduce costs.
  
* **Emerging trends:** The field is rapidly evolving.  Multimodal LLMs (e.g. GPT-4o) now handle text *and* images, broadening data types.  Agent-based “AutoResearch” frameworks have been proposed, where multiple LLM agents automate an entire research pipeline (literature review, hypothesis, methods, writing, etc.).  Retrieval-augmented models and real-time data connectivity are also advancing, enabling LLMs to access up-to-date databases.  Researchers are exploring these trends while carefully addressing the new ethical and methodological challenges they bring.



### The Future of Academic Research (Or Human Beings) with AI 
---

> `Jeffery Hinton`: *Human Beings will be the second most intelligent species on the Earth*.
    
* *AI-powered* machines will more than just a tool for humans, they will become *co-workers* and *collaborators* in the research process.
  
* We still not know what's the consequence of this change, but it will be profound.
  
* **AI will change the nature of academic research**: It's important to notice that many researchers will be replaced, though not all. Seek to make long-run impact!

* Focus on important research questions that require human intuition, creativity, and critical thinking. 
  
   * Embrace AI as a tool to enhance research efficiency, but maintain human oversight and ethical standards.
  
   * Stay updated on AI advancements and adapt to new tools and methodologies as they emerge.




<div align="center">
<b>Good Luck to Us!</b> <br>
<img src="images/HM_social_system.png" alt="Human-Machine Social Systems" width="300"/>
</div>


